File: October 2022/Day 25/C++ Code/Maximum of all subarrays of size k.cpp
//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function template for C++

class Solution
{
  public:
    //Function to find maximum of each subarray of size k.
    vector <int> max_of_subarrays(int *arr, int n, int k)
    {
        // your code here
        deque<pair<int,int>> dq;
        vector<int> ans;
        int j=0;
        while(j<n) {
            while(dq.size()>0 && dq.back().second<=arr[j]) {
                dq.pop_back();
            }
            dq.push_back(make_pair(j,arr[j]));
            if(j-dq.front().first==k) {
                dq.pop_front();
            }
            if(j>=k-1) ans.push_back(dq.front().second);
            j++;
        }
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
	
	int t;
	cin >> t;
	
	while(t--){
	    
	    int n, k;
	    cin >> n >> k;
	    
	    int arr[n];
	    for(int i = 0;i<n;i++) 
	        cin >> arr[i];
	    Solution ob;
	    vector <int> res = ob.max_of_subarrays(arr, n, k);
	    for (int i = 0; i < res.size (); i++) 
	        cout << res[i] << " ";
	    cout << endl;
	    
	}
	
	return 0;
}
// } Driver Code Ends

File: Medium/Maximum of all subarrays of size k/maximum-of-all-subarrays-of-size-k.java
//{ Driver Code Starts
//Initial template for JAVA

import java.util.*;
import java.io.*;
import java.lang.*;

class Main
{
    public static void main(String args[])
    {
        //taking input using class Scanner
        Scanner sc = new Scanner(System.in);
        
        //taking total count of testcases
        int t = sc.nextInt();
        
        
        
        while(t-- > 0)
        {
            //taking total number of elements
            int n = sc.nextInt();
            
            //taking size of subArrays 
            int k = sc.nextInt();
            
            //Declaring and Intializing an array of size n
            int arr[] = new int[n];
            
            //adding all the elements to the array 
            for(int i = 0; i <n; i++)
            {
                arr[i] =sc.nextInt();
            }
            
            //Calling the method max_of_subarrays of class solve
            //and storing the result in an ArrayList
            ArrayList <Integer> res = new Solution().max_of_subarrays(arr, n, k);
            
            //printing the elements of the ArrayList
            for (int i = 0; i < res.size(); i++)
                System.out.print (res.get (i) + " ");
            System.out.println();
        }
    }
}
// } Driver Code Ends


//User function template for JAVA

class Solution
{
    static class Pair{
        int idx;
        int val;
        Pair(int idx,int val) {
            this.idx=idx;
            this.val=val;
        }
    }
    
    //Function to find maximum of each subarray of size k.
    static ArrayList <Integer> max_of_subarrays(int arr[], int n, int k)
    {
        // Your code here
        
        ArrayDeque<Pair> dq = new ArrayDeque<>();
        ArrayList<Integer> ans = new ArrayList<>();
        int j=0;
        while(j<n) {
            while(dq.size()>0 && dq.getLast().val<=arr[j]) {
                dq.removeLast();
            }
            dq.add(new Pair(j,arr[j]));
            if(j-dq.getFirst().idx==k) {
                dq.removeFirst();
            }
            if(j>=k-1) ans.add(dq.peek().val);
            j++;
        }
        return ans;
        
    }
}